### PR TITLE
adding testing

### PR DESCRIPTION
**Pull request by talamyaseen**

### CODE REVIEW

File: WeatherBots.Tests/InputReaderTests.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            result.Should().BeNull();
+        }
+
+        [Fact]
+        public void ReadDocument_ShouldReturnNull_ForWhitespaceOnly()
+        {
+            Console.SetIn(new StringReader("   \n"));
+            var result = InputReader.ReadDocument();
+            result.Should().BeNull();
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/InputReaderTests.cs:25]: For these kinds of cases, it is preferred to use `[Theory]` here.

The only thing that's different between the two methods are the data, but both should execute the same behavior and output the same result.

Plus, these two cases are testing the same lines of code, and the same business rule, which is: Should return null for different versions of empty string (white spaces, new line, null strings).

So, when u see yourself:
_Writing the same test but with different variations of the input, AND, you're testing the same lines of code or the same business rule, opt to use `[Theory]` instead of `[Fact]`, and pass in the string that's used for testing._

File: WeatherBots.Tests/InputReaderTests.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        [Fact]
+        public void ReadDocument_ShouldHandleNestedXml()
+        {
+            var input = "<root>\n<parent>\n<child>value</child>\n</parent>\n</root>";
+            Console.SetIn(new StringReader(input));
+            var doc = InputReader.ReadDocument();
+            doc.Should().Contain("<root>");
+            doc.Should().Contain("<parent>");
+            doc.Should().Contain("<child>value</child>");
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/InputReaderTests.cs:98]: For these cases, I don't think they need separate tests, and we can use the ones above that handle reading correct JSON and XML. The only difference between these and those are the string that's being read, and this is a perfect example of doing a `[Theory]`.

File: WeatherBots.Tests/InputReaderTests.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+using Xunit;
+
+namespace WeatherBots.Tests
+{
+    public class InputReaderTests
+    {
+        [Fact]
+        public void ReadDocument_ShouldReturnNull_ForEmptyInput()
+        {
+            Console.SetIn(new StringReader("\n"));
```
Comment (SALAHT4N) [inline WeatherBots.Tests/InputReaderTests.cs:14]: I'm not sure if the courses mentioned this, so I'll mention it here just in case.

Currently, the `InputReader` method is using Console.ReadLine directly in the code, which is considered a static method call and that can make our code coupled to other pieces of code.

What I mean by coupled is: we can't reuse our code without using `Console` static methods.

The design that might be better is to create an interface nameed `IReader` and inject it into the class. And we can have `ConsoleReader` class that implements the interface and inject it into the Input reader. 

This can make out code call the IReader instead of directly calling `Console.ReadLine()`. which will make it decoupled from Console static methods. (Notice how this is basically Open-closed principal. And notice how after doing this, it will be a candidate to use Mocking here)

File: WeatherBots.Tests/InputReaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+using FluentAssertions;
+using WeatherBots.Utils;
+using Xunit;
+
+namespace WeatherBots.Tests
+{
+    public class InputReaderTests
+    {
+        [Fact]
+        public void ReadDocument_ShouldReturnNull_ForEmptyInput()
```
Comment (SALAHT4N) [inline WeatherBots.Tests/InputReaderTests.cs:12]: You might have missed this in this class, because I saw other test classes not having this naming issue, but test methods' names should follow this format: 
```
methodUnderTest_ScenarioWeAreTesting_ExpectedBehaviorOrResult()
```

File: WeatherBots.Tests/ObserversAndBotsTests.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+
+            mockObs.Verify(x => x.Update(It.IsAny<WeatherData>()), Times.Once);
+
+
+            station.Detach(mockObs.Object);
+            station.Publish(data);
+            mockObs.Verify(x => x.Update(It.IsAny<WeatherData>()), Times.Once, "should not be called after detach");
+        }
+
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:33]: This method name tells that it is testing all of the three methods in WeatherStation, which is considered a lot of methods in a single test to be tested.

By looking at the test code, we're making sure that publish only calls the `Update` on the attached bots only.

So, we can re-write this in a way that we define two mock bots, we only one of them, and verify that `Publish` only invoked one of the bot's `Update` method.

File: WeatherBots.Tests/ObserversAndBotsTests.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            var station = new WeatherStation();
+            var mockObs = new Mock<IWeatherObserver>();
+            var data = new WeatherData { Location = "L", Temperature = 1, Humidity = 2 };
+
+
+            station.Attach(mockObs.Object);
+            station.Publish(data);
+
+
+            mockObs.Verify(x => x.Update(It.IsAny<WeatherData>()), Times.Once);
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:26]: NIT: it's more accurate to verify that `Update` was called with the same `data` object defined above, and not any `It.IsAny<WeatherData>()` to be more precise.

File: WeatherBots.Tests/InputReaderTests.cs (hunk 3 above)
Comment (SALAHT4N) [reply WeatherBots.Tests/InputReaderTests.cs:14]: Now, for the sake of Console, you mocked it's behavior really well by controlling the input and output sources. Nice work!

But, if the methods calls `DateTime.Now` and there's some logic depending on the value of that time, the test will turn into a flaky test; meaning that it may fail or pass depending on what hour we are running the test.

example:
```
public void method1() 
{
    var date = DateTime.Now;
    if (date.Day == Days.Tuesday) { //do some logic }    
}
```

We can't mock this method and DateTime.Now value will be the value when we are running the test.

File: WeatherBots.Tests/ObserversAndBotsTests.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+            var sw = new StringWriter();
+            Console.SetOut(sw);
+
+
+            bot.Update(data);
+
+
+            sw.ToString().Should().Contain("RainBot activated!");
+            sw.ToString().Should().Contain("raining");
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:51]: It's better to use `cfg.Message` instead or rewriting the same literal value here, because, we might miss a letter and cause the test to fail.

File: WeatherBots.Tests/ObserversAndBotsTests.cs
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:1]: For this test class, I think it's testing too many classes, and we might consider splitting the test classes to target each class individually.
Since the method is testing the WeatherStation and all of the bots

File: WeatherBots.Tests/ParserAndParsersTests.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        [Fact]
+        public void XmlWeatherParser_ShouldReturnFalseForInvalidXml()
+        {
+            var parser = new XmlWeatherParser();
+            var ok = parser.TryParse("<WeatherData><Invalid></WeatherData>", out var data);
+            ok.Should().BeFalse();
+            data.Should().BeNull();
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ParserAndParsersTests.cs:52]: Some methods here are repeated from the Json test class before.

File: WeatherBots.Tests/ParserAndParsersTests.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                data = null;
+                if (input == "THISTESTPARSERINPUT")
+                {
+                    data = new WeatherData { Location = "test", Temperature = 0, Humidity = 0 };
+                    return true;
+                }
+                return false;
+            }
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ParserAndParsersTests.cs:102]: Nice!

Notice how this is basically mocking, but it allows us to have more control of the implementation, while mocking mainly controls what is outputted from which input.

File: WeatherBots.Tests/ParserAndParsersTests.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            var parser = new TestParser();
+            ParserFactory.RegisterParser(parser);
+
+            // Act
+            var ok = ParserFactory.TryParse("THISTESTPARSERINPUT", out var data);
+
+            // Assert
+            ok.Should().BeTrue();
+            data!.Location.Should().Be("test");
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ParserAndParsersTests.cs:87]: It could be helpful here to add a test case where we're trying to parse this same string but it returns false because the parser isn't registered.

File: WeatherBots.Tests/ParserAndParsersTests.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+
+        [Fact]
+        public void ParserFactory_TryParse_TriesAllRegisteredParsers()
+        {
+            var json = "{ \"Location\": \"Z\", \"Temperature\": 2, \"Humidity\": 3 }";
+            var ok = ParserFactory.TryParse(json, out var data);
+            ok.Should().BeTrue();
+            data.Should().NotBeNull();
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ParserAndParsersTests.cs:71]: This case doesn't really test what it's intending to test, the json parser is the first one and if this test passes, we're still not sure if the method actually loops over all of the parsers.

An approach that works could look like this:
```
[Fact]
public void TryParse_CallsAllParsersUntilOneSucceeds()
{
    var mock1 = new Mock<IWeatherParser>();
    var mock2 = new Mock<IWeatherParser>();
    var mock3 = new Mock<IWeatherParser>();

    mock1.Setup(m => m.TryParse(It.IsAny<string>(), out It.Ref<WeatherData?>.IsAny))
         .Returns(false);
    mock2.Setup(m => m.TryParse(It.IsAny<string>(), out It.Ref<WeatherData?>.IsAny))
         .Returns(true);

    Weather.Parsers = new[] { mock1.Object, mock2.Object, mock3.Object };

    Weather.TryParse("input", out _);

    mock1.Verify(m => m.TryParse(It.IsAny<string>(), out It.Ref<WeatherData?>.IsAny), Times.Once);
    mock2.Verify(m => m.TryParse(It.IsAny<string>(), out It.Ref<WeatherData?>.IsAny), Times.Once);
    mock3.Verify(m => m.TryParse(It.IsAny<string>(), out It.Ref<WeatherData?>.IsAny), Times.Never);
}
```

Here, it's visible from the test that the method executes a loop through the available parsers until it find a suitable one.

File: WeatherBots.Tests/WeatherStationTests.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            var station = new WeatherStation();
+            var mockObserver = new Mock<IWeatherObserver>();
+
+            station.Attach(mockObserver.Object);
+
+            var weather = new WeatherData { Location = "X", Temperature = 20, Humidity = 50 };
+            station.Publish(weather);
+
+            mockObserver.Verify(o => o.Update(weather), Times.Once);
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/WeatherStationTests.cs:23]: This is great!
NIT: add another mockObserver and don't attach it, just for the test method to be extra sure that we're calling `Update` of the attached bots ONLY. and not all of them.

It's not as clear here since we only have one bot, so it can mean we're calling Update on every bot, or just this one.

File: WeatherBots.Tests/ObserversAndBotsTests.cs (hunk 5 above)
Comment (SALAHT4N) [reply WeatherBots.Tests/ObserversAndBotsTests.cs:33]: EDIT: After reading the `WeatherStationTests.cs`, I think this test method was used to test the complete flow in using these classes, and this is indeed useful in some scenarios. Good job!.

File: WeatherBots.Tests/WeatherStationTests.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+            station.Attach(mockObserver.Object);
+
+            station.Detach(mockObserver.Object);
+
+            var weather = new WeatherData();
+            station.Publish(weather);
+
+            mockObserver.Verify(o => o.Update(It.IsAny<WeatherData>()), Times.Never);
+        }
+    }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/WeatherStationTests.cs:39]: This class isn't currently testing this if statement in the attach and detach methods: 
```
if (!_observers.Contains(observer))
```

I know this logic is simple, but a test method would useful here to protect the code from future breaking from future code changes.

We can test it by attaching the same bot mock twice, and make sure that the `Update` is only called once and not twice.

File: WeatherBots.Tests/ConfigLoaderTests.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        public void Load_WithValidJsonFile_ReturnsPopulatedBotConfigRoot()
+        {
+            // Arrange
+            var json = "{\"RainBot\": { \"enabled\": true, \"humidityThreshold\": 70, \"message\": \"It looks like it's about to pour down!\" }," +
+                " \"SunBot\": { \"enabled\": true, \"temperatureThreshold\": 30, \"message\": \"Scorching!\" }}";
+            var tmp = Path.GetTempFileName();
+            File.WriteAllText(tmp, json);
+
+            // Act
+            var root = ConfigLoader.Load(tmp);
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ConfigLoaderTests.cs:23]: *Implementation of Load*

This is also another example of where directly calling `File` methods in the Load method makes the code coupled, and we have to create a file and delete it in the test since the logic for files is inside the `Load` method and cannot be separated.

Consider wrapping the calls to File in a class and expose an interface to call them which makes the code more testable.

File: WeatherBots.Tests/ConfigLoaderTests.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            var tmp = Path.GetTempFileName();
+            File.WriteAllText(tmp, string.Empty);
+
+            Action act = () => ConfigLoader.Load(tmp);
+
+            act.Should().Throw<JsonException>();
+            File.Delete(tmp);
+        }
+
+    }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ConfigLoaderTests.cs:63]: NIT: These can be grouped into one `[Theory]` where they all represent an invalid text that results into an exception.

File: WeatherBots.Tests/CompletenessCheckerTests.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+        [Fact]
+        public void XmlCompletenessChecker_ShouldRemainIncompleteWithoutClosingTag()
+        {
+            var checker = new XmlCompletenessChecker();
+            checker.Consume("<root>");
+            checker.IsComplete.Should().BeFalse();
+        }
+
+    }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/CompletenessCheckerTests.cs:75]: I think all of this test class could be converted into two big `[Theory]`, one for XML and one for JSON, where each inline test case passes a list of strings.

Currently, all of the methods are trying to test the same business rule, but they all serve as different variations to make sure the method is robust and can handle edge cases, and this is good.

Also, a good candidate for `[Theory]` is when the method works as an algorithm with many test cases: Like Leetcode questions.
The code increments or decrements a counter based on the character is reading, so it's like an algorithm.

File: WeatherBots.Tests/CompletenessCheckerTests.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        [Theory]
+        [InlineData("{", typeof(JsonCompletenessChecker))]
+        [InlineData("[", typeof(JsonCompletenessChecker))]
+        [InlineData("<root>", typeof(XmlCompletenessChecker))]
+        public void Factory_ReturnsCorrectCheckerType(string firstLine, Type expected)
+        {
+            var checker = CompletenessCheckerFactory.CreateChecker(firstLine);
+            checker.Should().BeOfType(expected);
+        }
```
Comment (SALAHT4N) [inline WeatherBots.Tests/CompletenessCheckerTests.cs:18]: Nice!

**Review by SALAHT4N — CHANGES_REQUESTED** [review]

**Comment by SALAHT4N** [comment]
### Part 2

File: WeatherBots.Tests/ObserversAndBotsTests.cs [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+            mockObs.Verify(x => x.Update(It.IsAny<WeatherData>()), Times.Once, "should not be called after detach");
+        }
+
+
+        [Fact]
+        public void RainBot_WhenHumidityAboveThreshold_WritesToConsole()
+        {
+            var cfg = new Config.BotConfig { Enabled = true, HumidityThreshold = 50, Message = "raining" };
+            var bot = new RainBot(cfg);
+            var data = new WeatherData { Location = "x", Temperature = 10, Humidity = 60 };
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:40]: When u find yourself using values like "x" or just dummy strings, it's a good sign to utilize AutoFixture which can generate random data in the Arrange section of test methods.

We use fixture when we don't care about the data that's created and we just need data to satisfy method parameters. But sometimes we want specific fields to have a certain value, where we can use fixture builders to build the object how we want.

File: WeatherBots.Tests/ObserversAndBotsTests.cs [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+            var bot = new SnowBot(cfg);
+            var data = new WeatherData { Location = "X", Temperature = -5, Humidity = 50 };
+
+            var sw = new StringWriter();
+            Console.SetOut(sw);
+
+            bot.Update(data);
+
+            sw.ToString().Should().Contain("SnowBot activated!");
+            sw.ToString().Should().Contain("snowy");
```
Comment (SALAHT4N) [inline WeatherBots.Tests/ObserversAndBotsTests.cs:86]: When doing multiple assertions like this, it's better to wrap them in an `AssertionScope` to make out lives easier.

If the first assertion fails, the method won't execute the second assertion, so we might go and fix the first one, only to find out that the second assertions also fails.

Imagine if 4 or 5 assertions, that would make it not fun to write tests.
If we wrap them in an assertion scope, then all asserts will run and we will get the failure of each one on the same run.

File: WeatherBots.Tests/WeatherStationTests.cs [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        [Fact]
+        public void Attach_And_Publish_ShouldCallObserver()
+        {
+            var station = new WeatherStation();
+            var mockObserver = new Mock<IWeatherObserver>();
+
+            station.Attach(mockObserver.Object);
+
+            var weather = new WeatherData { Location = "X", Temperature = 20, Humidity = 50 };
```
Comment (SALAHT4N) [inline WeatherBots.Tests/WeatherStationTests.cs:19]: use fixture, we don't care about the data, we just want an object of **_any_** data.

File: WeatherBots.Tests/InputReaderTests.cs [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+            var input = "unsupported format line\n";
+            Console.SetIn(new StringReader(input));
+            Action act = () => InputReader.ReadDocument();
+            act.Should().Throw<NotSupportedException>();
+        }
+
+        [Fact]
+        public void ReadDocument_ShouldHandleNestedJson()
+        {
+            var input = "{\n\"Outer\": { \"Inner\": { \"Value\": 123 } }\n}";
```
Comment (SALAHT4N) [inline WeatherBots.Tests/InputReaderTests.cs:81]: Make sure to put down the comments: 

```
// Arrange
...
// Act
...
// Assert
```
